src/frontegg-client/src/client.rs

```diff
@@ -0,0 +1,207 @@
+    pub expires_in: i64,
+    pub refresh_token: String,
+}
+
+/// An API client for Frontegg.
+///
+/// The API client is designed to be wrapped in an [`Arc`] and used from
+/// multiple threads simultaneously. A successful authentication response is
+/// shared by all threads.
+///
+/// [`Arc`]: std::sync::Arc
+pub struct Client {
+    pub inner: reqwest::Client,
+    pub app_password: AppPassword,
+    pub endpoint: Url,
+    pub auth: Mutex<Option<Auth>>,
+}
+
+impl Client {
+    /// Creates a new `Client` from its required configuration parameters.
+    pub fn new(config: ClientConfig) -> Client {
+        ClientBuilder::default().build(config)
+    }
+
+    /// Creates a builder for a `Client` that allows for customization of
+    /// optional parameters.
+    pub fn builder() -> ClientBuilder {
+        ClientBuilder::default()
+    }
+
+    /// Builds a request towards the `Client`'s endpoint
+    fn build_request<P>(&self, method: Method, path: P) -> RequestBuilder
+    where
+        P: IntoIterator,
+        P::Item: AsRef<str>,
+    {
+        let mut url = self.endpoint.clone();
+        url.path_segments_mut()
+            .expect("builder validated URL can be a base")
+            .clear()
+            .extend(path);
+        self.inner.request(method, url)
+    }
+
+    /// Sends a requests and adds the authorization bearer token.
+    async fn send_request<T>(&self, req: RequestBuilder) -> Result<T, ErrorExtended>
+    where
+        T: DeserializeOwned,
+    {
+        let token = self.auth().await?.token;
+        let req = req.bearer_auth(token);
+        self.send_unauthenticated_request(req).await
+    }
+
+    async fn send_unauthenticated_request<T>(&self, req: RequestBuilder) -> Result<T, ErrorExtended>
+    where
+        T: DeserializeOwned,
+    {
+        #[derive(Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct ErrorResponse {
+            #[serde(default)]
+            message: Option<String>,
+            #[serde(default)]
+            errors: Vec<String>,
+        }
+
+        let res = req.send().await?;
+        let status_code = res.status();
+        if status_code.is_success() {
+            Ok(res.json().await?)
+        } else {
+            match res.json::<ErrorResponse>().await {
+                Ok(e) => {
+                    let mut messages = e.errors;
+                    messages.extend(e.message);
+                    Err(ErrorExtended::Api(ApiError {
+                        status_code,
+                        messages,
+                    }))
+                }
+                Err(_) => Err(ErrorExtended::Api(ApiError {
+                    status_code,
+                    messages: vec!["unable to decode error details".into()],
+                })),
+            }
+        }
+    }
+
+    /// Authenticates with the server, if not already authenticated,
+    /// and returns the authentication token.
+    pub async fn auth(&self) -> Result<Auth, ErrorExtended> {
```

Comment:
Does this need to be pub?

Reply:
Yes. I'll use it in the future to get the token from other crates, like `cloud-api`.
I've changed it to `pub async fn auth(&self) -> Result<String, ErrorExtended>`, to return only the token as a String.